instantout/reservation/store.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// SQLStore manages the reservations in the database.
+type SQLStore struct {
+	baseDb BaseDB
+
+	nowTime func() time.Time
+}
+
+// NewSQLStore creates a new SQLStore.
+func NewSQLStore(db BaseDB) *SQLStore {
+	nowTime := time.Now
+
+	return &SQLStore{
+		baseDb:  db,
+		nowTime: nowTime,
+	}
+}
+
+// CreateReservation stores the reservation in the database.
+func (r *SQLStore) CreateReservation(ctx context.Context,
+	reservation *Reservation) error {
+
+	args := sqlc.CreateReservationParams{
+		ReservationID:    reservation.ID[:],
+		ClientPubkey:     reservation.ClientPubkey.SerializeCompressed(),
+		ServerPubkey:     reservation.ServerPubkey.SerializeCompressed(),
+		Expiry:           int32(reservation.Expiry),
+		Value:            int64(reservation.Value),
+		ClientKeyFamily:  int32(reservation.KeyLocator.Family),
+		ClientKeyIndex:   int32(reservation.KeyLocator.Index),
+		InitiationHeight: reservation.InitiationHeight,
+	}
+
+	updateArgs := sqlc.InsertReservationUpdateParams{
+		ReservationID:   reservation.ID[:],
+		UpdateTimestamp: r.nowTime().UTC(),
+		UpdateState:     string(reservation.State),
+	}
+
+	return r.baseDb.ExecTx(ctx, &loopdb.SqliteTxOptions{},
+		func(q *sqlc.Queries) error {
+			err := q.CreateReservation(ctx, args)
+			if err != nil {
+				return err
+			}
+
+			return q.InsertReservationUpdate(ctx, updateArgs)
+		})
+}
+
+// UpdateReservation updates the reservation in the database.
+func (r *SQLStore) UpdateReservation(ctx context.Context,
+	reservation *Reservation) error {
+
+	var txHash []byte
+	var outIndex sql.NullInt32
+	if reservation.Outpoint != nil {
+		txHash = reservation.Outpoint.Hash[:]
+		outIndex = sql.NullInt32{
+			Int32: int32(reservation.Outpoint.Index),
+			Valid: true,
+		}
+	}
+
+	insertUpdateArgs := sqlc.InsertReservationUpdateParams{
+		ReservationID:   reservation.ID[:],
+		UpdateTimestamp: r.nowTime().UTC(),
+		UpdateState:     string(reservation.State),
+	}
+
+	updateArgs := sqlc.UpdateReservationParams{
+		ReservationID: reservation.ID[:],
+		TxHash:        txHash,
+		OutIndex:      outIndex,
+		ConfirmationHeight: marshalSqlNullInt32(
+			int32(reservation.ConfirmationHeight),
+		),
+	}
+
+	return r.baseDb.ExecTx(ctx, &loopdb.SqliteTxOptions{},
+		func(q *sqlc.Queries) error {
+			err := q.UpdateReservation(ctx, updateArgs)
+			if err != nil {
+				return err
+			}
+
+			return q.InsertReservationUpdate(ctx, insertUpdateArgs)
+		})
+}
+
+// GetReservation retrieves the reservation from the database.
+func (r *SQLStore) GetReservation(ctx context.Context,
+	reservationId ID) (*Reservation, error) {
+
+	var reservation *Reservation
+	err := r.baseDb.ExecTx(ctx, loopdb.NewSqlReadOpts(),
+		func(q *sqlc.Queries) error {
+			var err error
+			reservationRow, err := q.GetReservation(
+				ctx, reservationId[:],
+			)
+			if err != nil {
+				return err
+			}
+
+			reservationUpdates, err := q.GetReservationUpdates(
+				ctx, reservationId[:],
+			)
+			if err != nil {
+				return err
+			}
```

Comment:
nit: maybe check that there's at least one update. Not strictly needed but better to be safe.